src/waitress/channel.py

```diff
@@ -176,15 +186,14 @@ def received(self, data):
                 request.expect_continue = False
 
                 if not self.sent_continue:
-                    # there's no current task, so we don't need to try to
-                    # lock the outbuf to append to it.
                     outbuf_payload = b"HTTP/1.1 100 Continue\r\n\r\n"
                     num_bytes = len(outbuf_payload)
-                    self.outbufs[-1].append(outbuf_payload)
-                    self.current_outbuf_count += num_bytes
-                    self.total_outbufs_len += num_bytes
-                    self.sent_continue = True
-                    self._flush_some()
+                    with self.outbuf_lock:
+                        self.outbufs[-1].append(outbuf_payload)
+                        self.current_outbuf_count += num_bytes
+                        self.total_outbufs_len += num_bytes
+                        self.sent_continue = True
+                        self._flush_some()
```

Comment:
I think the problem with this is that we do not want to send responses out of order. Now that we can be reading a new request while processing a current request - this needs to be done in such a way that it does not send data on the outbufs until the current request is completed with its processing.

This may require checking if a current request is executing, and if so then buffering this data somewhere and flushing it when the request is complete. Or maybe more simply this entire branch of logic that handles 100 Continue needs to be moved into a task itself instead of being handled here.

I think the problem I'm identifying here is real but I'm not confident in my proposed solutions.

Reply:
Simplest solution may be to stop processing lookahead requests and don't resume until the current request is completed when we hit an expect-continue.

Reply:
A very good point. This work really allows me to understand more about the HTTP protocol than I ever thought I would need. I was not aware of the meaning and mechanics of `100 Continue` before.

I guess I will first add a test that pushes a request that waits a second before returning a defined result and immediately afterwards pushes a request with expect-continue, checking that the responses come in the right order (which will fail now).

Your suggested solution should work, although I think it would hinder the detection of client disconnects if an expect-continue occurs. Maybe I can think of something better, else this solution will have to do.

Reply:
I added a (currently failing) test to check for this. The thing is - I am pretty sure the situation is not covered on the current master branch either. At least not if `handle_read` finds one regular request and one with expect-continue that are both read with the same `recv` call. But now that we have identified the potential problem, we should fix it.

Reply:
hmm you may be right that it's broken right now if it receives one full request followed by an expect request in the same recv call.

It's worth noting also that waitress is caching `self.sent_continue` for the lifetime of the entire connection, so there's an expectation that it won't receive more than one expect request on the same connection and I'm not sure how valid that assumption is.

Reply:
It's an invalid assumption according to the RFC's, but no web browser (and I would argue almost all clients) don't actually support sending Expect Continue's in the first place, which is why nobody has yelled about it before.